ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/API.hs

```diff
@@ -437,7 +439,7 @@ addBlockWaitWrittenToDisk chainDB punish blk = do
 
 -- | Add a block synchronously: wait until the block has been processed (see
 -- 'blockProcessed'). The new tip of the ChainDB is returned.
-addBlock :: IOLike m => ChainDB m blk -> InvalidBlockPunishment m -> blk -> m (Point blk)
+addBlock :: IOLike m => ChainDB m blk -> InvalidBlockPunishment m -> blk -> m (Maybe (Point blk))
```

Comment:
1) `grep` hasn't yet found a use of this function that is obviously part of the implementation. I'm suspecting it's only for tests.

2) Therefore, I'm less worried about muddying its type with this partiality that should only happen when the whole node is shutting down. However, if we ever do want to remove the `Maybe` you're adding here, since its semantics is "return once the block is processed", I think it'd be fine to raise an exception instead of ever returning `Nothing`.

Reply:
The definition `addBlock_ x y z = void $ addBlock x y z` below forces this question a little more:

- Should `addBlock_` return quietly when `addBlock` returned `Nothing`?

- Or should it instead raise that hypothetical exception I had proposed in the previous message?

(I'm also not definitely seeing `addBlock_` used in the actual implementation.)

Reply:
`addBlock_` is used by [`initNodeKernel`](https://input-output-hk.github.io/ouroboros-network/ouroboros-consensus/Ouroboros-Consensus-NodeKernel.html#v:initNodeKernel) via [`fromFull` ](https://input-output-hk.github.io/ouroboros-network/ouroboros-consensus/Ouroboros-Consensus-Storage-ChainDB-Init.html#v:fromFull).  So these function should not be partial.

Reply:
This message is just exhaustively checking for uses.

`ouroboros-consensus*` defines four `addBlock` functions:

- `addBlock`          380 `ouroboros-network/ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/Model.hs` signature
- `addBlock`          440 `ouroboros-network/ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/API.hs` signature
- `addBlock`           24 `ouroboros-network/ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Init.hs` selector
- `addBlock`           63 `ouroboros-network/ouroboros-consensus/src/Ouroboros/Consensus/Storage/VolatileDB/Impl/FileInfo.hs` signature

The second one is our topic here and neighbors the one and only `addBlock_` wrapper.

The same name is also defined in some `ouroboros-network*` packages.

`git grep -w -e addBlock origin/master -- 'ouroboros*hs'` reveals these uses:

- `API.addBlock_` used in `API.fromChain`, which seems to be dead code (maybe downstream repo?)
- `API.addBlock_` used in `Init.fromFull`, see below.
- `API.addBlock_` used in `Test.Ouroboros.Storage.ChainDB.FollowerPromptness`
- `API.addBlock` used in `Test.Ouroboros.Storage.ChainDB.StateMachine`
- `API.addBlock` used in `Test.ThreadNet.Network`
- `API.addBlock` used to define `API.toChain`, which is only used in `Test.ThreadNet.Network`.

- `FileInfo.addBlock` used in `Ouroboros.Consensus.Storage.VolatileDB.Impl`

- `Init.addBlock` used in `nodeInitChainDB @ByronBlock`
- `Init.addBlock` used in `nodeInitChainDB @DualByronBlock`
- `Init.addBlock` composed in `nodeInitChainDB @HardForkBlock`

- `Model.addBlock` also used in `Test.Ouroboros.Storage.ChainDB.StateMachine`

- `Ouroboros.Network.MockChain.Chain.addBlock` used in `Test.Util.ChainUpdates`

- `Test.Ouroboros.Storage.ChainDB.Model{,.Test}` defines and uses its own `addBlock`

(There's a second `fromChain`, which is declared in `Ouroboros/Consensus/HeaderStateHistory.hs`, but it's unrelated and only used in `Test.Consensus.MiniProtocol.ChainSync.Client`.)

That confirms `Init.fromFull` as the only interesting one, as you had noted.

It's used to initialize the node in `NodeKernel.hs`; in this call:

https://github.com/input-output-hk/ouroboros-network/blob/72863b0fc78abdc2b8e29f0dda96c06da3dd11d0/ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs#L138-L143

`git grep` indicates that `initChainDB` is only defined as `nodeInitChainDB`. So the use-def chain looks like this: `initChainDB`, `nodeInitChainDB`, `Init.addBlock` (for Byron types), `Init.fromFull`, `API.addBlock_`, `API.addBlock`.

Reply:
Looking more closely at `nodeInitChainDB @ByronBlock` --- which is ultimately the only interesting transitive use of `API.addBlock` (see previous message) --- we see it's merely adding the slot 0 Epoch Boundary Block.

https://github.com/input-output-hk/ouroboros-network/blob/72863b0fc78abdc2b8e29f0dda96c06da3dd11d0/ouroboros-consensus-byron/src/Ouroboros/Consensus/Byron/Node.hs#L273-L282

There's no way to recover, if that fails.

---

So: the only real use of `addBlock` in the system has no expectation of failure and no useful way to recover if it did fail.

Reply:
Because there is a failure case in the actual implementation, my preference is to keep the `Maybe` layer you've added here (but as a custom `Maybe`-isomorph with clearly named constructors), and to propagate it all the way to the problematic call-site.

EG we'd need to update at least `Init.addBlock` to return `Maybe ()` instead of `()`. Perhaps also doing that for `nodeInitChainDB`.

But I'm not certain about this. I think I tend to reach for explicit error types more than most. So maybe a `UnrecoverableFailureDuringNodeInit` would be a fine exception, and it could be wired into the definition of `Init.addBlock :: ... -> m ()`?

What do you think, @coot